detoks/src/main/java/nl/tudelft/trustchain/detoks/gossiper/NetworkSizeGossiper.kt

```diff
@@ -0,0 +1,109 @@
+package nl.tudelft.trustchain.detoks.gossiper
+
+import android.util.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import nl.tudelft.ipv8.Peer
+import nl.tudelft.ipv8.android.IPv8Android
+import nl.tudelft.trustchain.detoks.DeToksCommunity
+import kotlin.random.Random.Default.nextDouble
+
+class NetworkSizeGossiper(
+    override val delay: Long,
+    override val peers: Int,
+    private val leaders: Int
+) : Gossiper() {
+
+    private var firstCycle = true
+
+    override fun startGossip(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
+            while (coroutineScope.isActive) {
+                if (firstCycle) {
+                    firstCycle = false
+                    delay(System.currentTimeMillis() % delay)
+                }
+                gossip()
+                delay(delay)
+            }
+        }
+    }
+
+    override suspend fun gossip() {
+        val deToksCommunity = IPv8Android.getInstance().getOverlay<DeToksCommunity>()!!
+
+        if (leaderEstimates.isNotEmpty())
+            networkSizeEstimate = (1 / leaderEstimates.minOfOrNull { it.second }!!).toInt()
+
+        awaitingResponse.clear()
+
+        val chanceLeader = leaders / (networkSizeEstimate.toDouble())
+        Log.d(DeToksCommunity.LOGGING_TAG, "Chance to become leader: $chanceLeader, leaders: $leaders, networksize estimate: $networkSizeEstimate")
+        leaderEstimates = if (nextDouble() < chanceLeader)
+            listOf(Pair(deToksCommunity.myPeer.mid, 1.0))
+        else listOf()
```

Comment:
If I understand correctly this seems that there are no leaders, first of all shouldn't the list have type ? Also maybe some message to represent that there is no new leader